pkg/applicationserver/io/web/templates.go

```diff
@@ -0,0 +1,144 @@
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package web
+
+import (
+	"context"
+	"encoding/json"
+	"fmt"
+	"sync"
+
+	"go.thethings.network/lorawan-stack/pkg/fetch"
+	"go.thethings.network/lorawan-stack/pkg/ttnpb"
+)
+
+// TemplateStore contains the webhook templates.
+type TemplateStore struct {
+	Fetcher fetch.Interface
+
+	templates      sync.Map
+	templateIDs    *[]string
+	templateIDsMu  sync.Mutex
+	templateIDsErr error
+}
+
+// NewTemplateStore creates a new template store that is backed by the provided fetcher.
+func NewTemplateStore(fetcher fetch.Interface) (*TemplateStore, error) {
+	return &TemplateStore{
+		Fetcher: fetcher,
+	}, nil
+}
+
+// Get returns the template with the given identifiers.
+func (ts *TemplateStore) Get(ctx context.Context, req *ttnpb.GetApplicationWebhookTemplateRequest) (*ttnpb.ApplicationWebhookTemplate, error) {
+	template, err := ts.getTemplate(req.ApplicationWebhookTemplateIdentifiers)
+	if err != nil {
+		return nil, err
+	}
+	return applyWebhookTemplateFieldMask(nil, template, appendImplicitWebhookTemplatePaths(req.FieldMask.Paths...)...)
+}
+
+// List returns the available templates.
+func (ts *TemplateStore) List(ctx context.Context, req *ttnpb.ListApplicationWebhookTemplatesRequest) (*ttnpb.ApplicationWebhookTemplates, error) {
+	ids, err := ts.getAllTemplateIDs()
+	if err != nil {
+		return nil, err
+	}
+
+	var templates ttnpb.ApplicationWebhookTemplates
+	for _, id := range ids {
+		template, err := ts.getTemplate(ttnpb.ApplicationWebhookTemplateIdentifiers{
+			TemplateID: id,
+		})
+		if err != nil {
+			return nil, err
+		}
+
+		template, err = applyWebhookTemplateFieldMask(nil, template, appendImplicitWebhookTemplatePaths(req.FieldMask.Paths...)...)
+		if err != nil {
+			return nil, err
+		}
+
+		templates.Templates = append(templates.Templates, template)
+	}
+	return &templates, nil
+}
+
+type registeredTemplate struct {
+	t     *ttnpb.ApplicationWebhookTemplate
+	err   error
+	ready chan struct{}
+}
+
+func (ts *TemplateStore) getAllTemplateIDs() (ids []string, err error) {
+	ts.templateIDsMu.Lock()
+	defer ts.templateIDsMu.Unlock()
+	if ts.templateIDs != nil {
+		return *ts.templateIDs, ts.templateIDsErr
+	}
+	defer func() {
+		ts.templateIDs, ts.templateIDsErr = &ids, err
+	}()
+
+	data, err := ts.Fetcher.File("templates.json")
+	if err != nil {
+		return nil, err
+	}
+	err = json.Unmarshal(data, &ids)
+	if err != nil {
+		return nil, err
+	}
+	return ids, err
+}
+
+func (ts *TemplateStore) getTemplate(ids ttnpb.ApplicationWebhookTemplateIdentifiers) (t *ttnpb.ApplicationWebhookTemplate, err error) {
+	registeredI, ok := ts.templates.LoadOrStore(ids.TemplateID, &registeredTemplate{ready: make(chan struct{})})
+	registered := registeredI.(*registeredTemplate)
+	if ok {
+		<-registered.ready
+		return registered.t, registered.err
+	}
+	defer func() {
+		registered.t, registered.err = t, err
+		close(registered.ready)
+	}()
```

Comment:
What is the above for?

Reply:
Synchronization - allows only one of the `get` calls to update the map. Will remove along with the caching.

Reply:
Yeah this is all the fetcher's responsibility

Reply:
I don't think caching this is such a bad thing actually. Otherwise we'll be unmarshaling the same thing all the time.

Reply:
If we do the caching, I rather see it being consistent with the (TTL) caching that we have in `pkg/frequencyplans`.

Note that `pkg/devicerepository` doesn't do any caching but is subject to change a bit anyway (#263).

Ideally we have a YAML decoder fetcher on top of the generic data fetcher that does caching, and that we can reuse across FPs, DR and this.